astropy/utils/parsing.py

```diff
@@ -0,0 +1,133 @@
+# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Wrappers for PLY to provide thread safety.
+"""
+
+import contextlib
+import functools
+import re
+import os
+import threading
+
+
+TAB_HEADER = """# -*- coding: utf-8 -*-
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+# This file was automatically generated from ply. To re-generate this file,
+# remove it from this folder, then build astropy and run the tests in-place:
+#
+#   python setup.py build_ext --inplace
+#   pytest {package}
+#
+# You can then commit the changes to this file.
+
+"""
+_LOCK = threading.RLock()
+
+
+def _add_tab_header(filename, package):
+    with open(filename, 'r') as f:
+        contents = f.read()
+
+    with open(filename, 'w') as f:
+        f.write(TAB_HEADER.format(package=package))
+        f.write(contents)
+
+
+@contextlib.contextmanager
+def _patch_get_caller_module_dict(module):
+    """Temporarily replace the module's get_caller_module_dict.
+
+    This is a function inside ``ply.lex`` and ``ply.yacc`` (each has a copy)
+    that is used to retrieve the caller's local symbols. Here, we patch the
+    function to instead retrieve the grandparent's local symbols to account
+    for a wrapper layer.
+    """
+    original = module.get_caller_module_dict
+
+    @functools.wraps(original)
+    def wrapper(levels):
+        # Add 2, not 1, because the wrapper itself adds another level
+        return original(levels + 2)
+
+    module.get_caller_module_dict = wrapper
+    yield
+    module.get_caller_module_dict = original
+
+
+def lex(lextab, package, reflags=int(re.VERBOSE)):
+    """Create a lexer from local variables.
+
+    It automatically compiles the lexer in optimized mode, writing to
+    ``lextab`` in the same directory as the calling file.
+
+    This function is thread-safe. The returned lexer is *not* thread-safe, but
+    if it is used exclusively with a single parser returned by :func:`yacc`
+    then it will be safe.
+
+    The ``package`` is inserted into the instructions in the generated file that
+    explain which package to test to rebuild it.
+
+    It is only intended to work with lexers defined within the calling
+    function, rather than at class or module scope.
+    """
+    from astropy.extern.ply import lex
+
+    caller_file = lex.get_caller_module_dict(2)['__file__']
+    lextab_filename = os.path.join(os.path.dirname(caller_file), lextab + '.py')
+    with _LOCK:
+        lextab_exists = os.path.exists(lextab_filename)
+        with _patch_get_caller_module_dict(lex):
+            lexer = lex.lex(optimize=True, lextab=lextab,
+                            outputdir=os.path.dirname(caller_file),
+                            reflags=reflags)
+        if not lextab_exists:
+            _add_tab_header(lextab_filename, package)
```

Comment:
can we have a new test that covers this line? codecov complains

Reply:
Done (in theory - I'll need to double-check codecov when it finishes).

Reply:
Done